### PR TITLE
2D decomposition on write grid, bug fix in FV3GFS_IO.F90, and quad precision in stochastic physics

### LINKED ISSUE #1023: Add 2d decomposition on the write grid comp
## Description
UPP is updating their code to allow 2D decomposition in the input data domain. To enable the capability on the inline post, the 2D decomposition will be enabled on write grid component so that the output data arrived on write grid component will also be on the post 2D decomposition domain. Currently the data on write grid component is on 1D decomposition domain (belted domain) to  align with the POST decomposition. 

## Solution
Set up i and j direction decomposition on write grid component and the write grid with 2D decomposition.

## Alternatives
If applicable, add a description of any alternative solutions or features you've considered.

## Related to
PR will come soon.

### PR DESCRIPTION
**Pull request by junwang-noaa**
# PR Checklist

- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [ ] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.

- [ ] New or updated input data is required by this PR. If checked, please work with the code managers to update input data sets on all platforms.

## Instructions: All subsequent sections of text should be filled in as appropriate.

The information provided below allows the code managers to understand the changes relevant to this PR, whether those changes are in the ufs-weather-model repository or in a subcomponent repository. Ufs-weather-model code managers will use the information provided to add any applicable labels, assign reviewers and place it in the Commit Queue. Once the PR is in the Commit Queue, it is the PR owner's responsiblity to keep the PR up-to-date with the develop branch of ufs-weather-model. 

## Description

The 2D decomposition will be enabled on write grid component so that the output data arrived on write grid component will also be on 2D decomposition domain that 2D post can use.

This PR also includes 1) FV3atm PR https://github.com/NOAA-EMC/fv3atm/pull/468 from @MicroTed  2) stochastic physics PR https://github.com/noaa-psd/stochastic_physics/pull/53 from Dom.
 
The code changes have no impact on the current baseline results. Two new tests (global: control_2ddecomp; regional: regional_control_2ddecomp) are added. Those tests generate the same results as the corresponding control tests.  

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues must always be created before starting work on a PR branch!) 
- fixes https://github.com/ufs-community/ufs-weather-model/issues/1023

## Testing

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [x] wcoss_cray
- [x] wcoss_dell_p3
- [ ] opnReqTest for newly added/changed feature
- [x] CI (72baf88)

## Dependencies

If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).

- waiting on https://github.com/NOAA-EMC/fv3atm/pull/470
- waiting on https://github.com/noaa-psd/stochastic_physics/pull/53

### CODE REVIEW

**Review by DusanJovic-NOAA — APPROVED** [review]

File: tests/rt.conf [hunk 1]
```diff
@@ -33,6 +33,7 @@ RUN     | cpld_debug_p8
 COMPILE | -DAPP=ATM -DCCPP_SUITES=FV3_GFS_v16,FV3_GFS_v15_thompson_mynn,FV3_GFS_v15_thompson_mynn_RRTMGP,FV3_GFS_v16_p7_rrtmgp,FV3_GFS_v16_p8 -D32BIT=ON   |                | fv3 |
 RUN     | control                                                                                                                 |                                         | fv3 |
 RUN     | control_decomp                                                                                                          |                                         |     |
+RUN     | control_2ddecomp                                                                                                        |                                         |     |
```
Comment (WenMeng-NOAA) [inline tests/rt.conf]: @junwang-noaa Would we expect  post results from control and control_2ddecomp are identical? Same expectation for regional_control and regional_control_2ddecomp?

Comment (DusanJovic-NOAA) [reply tests/rt.conf]: I would rename this test to `control_2dpostdecomp`, to indicate that '2d' is post's or (write) decomposition, not model's decomposition (i.e. layout)

Comment (junwang-noaa) [reply tests/rt.conf]: @WenMeng-NOAA Yes,  control and control_2ddecomp generate same results. Please see the "CNTL_DIR" in the test definition below. The control_2ddecomp is comparing results with control test(CNTL_DIR=control). The same for the regional 2ddecomp test.
@DusanJovic-NOAA renamed.

**Review by WenMeng-NOAA — APPROVED** [review]

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: hera
Compiler: intel
Job: RT
[RT] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/833875455/20220131223017/ufs-weather-model
[RT] Error: Test control_2dwrtdecomp 018 failed in check_result failed 
[RT] Error: Test control_2dwrtdecomp 018 failed in run_test failed 
Please make changes and add the following label back: hera-intel-RT

**Comment by junwang-noaa** [comment]
> _**Automated RT Failure Notification**_ Machine: hera Compiler: intel Job: RT [RT] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/833875455/20220131223017/ufs-weather-model [RT] Error: Test control_2dwrtdecomp 018 failed in check_result failed [RT] Error: Test control_2dwrtdecomp 018 failed in run_test failed Please make changes and add the following label back: hera-intel-RT

The control test has iovr specified from previous PR #1025. The control_2dwrtdecomp needs to be synced with this change.

**Comment by junwang-noaa** [comment]
@MinsukJi-NOAA The CI test failed, would you please take a look? Thanks.

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: gaea
Compiler: intel
Job: RT
[RT] Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/833875455/20220201181506/ufs-weather-model
Please make changes and add the following label back: gaea-intel-RT

**Comment by BrianCurtis-NOAA** [comment]
> _**Automated RT Failure Notification**_ Machine: gaea Compiler: intel Job: RT [RT] Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/833875455/20220201181506/ufs-weather-model Please make changes and add the following label back: gaea-intel-RT

They've been messing with the F2 filesystem on Gaea today. Seems like things are still running but I'm not sure the AutoRT will be there to push the log at the end if all goes well. its rt_27146 FYI.

**Comment by junwang-noaa** [comment]
I saw all the RT tests passed under /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/833875455/20220201181506/ufs-weather-model/tests/log_gaea.intel. But the RegressionTests_gaea.intel.log log file is not updated with the rt test log files under that directory.

**Comment by BrianCurtis-NOAA** [comment]
> I saw all the RT tests passed under /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/833875455/20220201181506/ufs-weather-model/tests/log_gaea.intel. But the RegressionTests_gaea.intel.log log file is not updated with the rt test log files under that directory.

Yes I saw that too. I am guessing for consistency that we want to upload completed logs, so I added the label for Gaea back on. Maybe we can write a script (python or bash) that generates the log on its own and integrate it into rt.sh ?

**Review by climbfuji — APPROVED** [review]
I am only approving the stochastic physics submodule pointer update (hash is correct), I haven't had time to look at the other changes.

**Pull request merged**